File: my-folder/0045-jump-game-ii/solution.java
class Solution {
    public int jump(int[] nums) {
        int jump = 0;
        int curEnd = 0;
        int curFar = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            curFar = Math.max(curFar, i + nums[i]);
            if (i == curEnd) {
                jump++;
                curEnd = curFar;
            }
        }
        return jump;
    }
}

File: my-folder/0045-jump-game-ii/solution.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        # The main idea is based on greedy. 
        # Let's say the range of the current jump is [curBegin, curEnd], curFarthest is the farthest point that all points in [curBegin, curEnd] can reach. 
        # Once the current point reaches curEnd, then trigger another jump, and set the new curEnd with curFarthest, then keep the above steps, as the following:

        res = 0
        curEnd = 0
        curFar = 0

        for i in range(len(nums) - 1):
            curFar = max(curFar, i + nums[i])
            if i == curEnd:
                res += 1
                curEnd = curFar

        return res
        

